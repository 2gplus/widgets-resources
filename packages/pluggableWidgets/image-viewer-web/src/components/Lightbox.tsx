import classNames from "classnames";
import { createElement, ReactElement, useCallback } from "react";
import Modal, { RenderModalBackdropProps } from "react-overlays/Modal";

import closeIconSvg from "../assets/ic24-close.svg";

export interface LightboxProps {
    children: ReactElement;
    isOpen: boolean;
    onClose: () => void;
}

interface RenderBackdropWithCloseProps extends RenderModalBackdropProps {
    onClose: () => void;
}

/**
 * The `onClick` here is provided by the `onHide` from the Modal props. But unfortunately it only works
 * for the actual backdrop and not for other elements due to a `e.target !== e.currentTarget` check in
 * the library. Therefore, we extend it with an additional `onClose` prop for our close button.
 */
function BackdropWithClose({ ref, onClose, onClick }: RenderBackdropWithCloseProps): ReactElement {
    return (
        <div className="mx-image-viewer-lightbox-backdrop" ref={ref} onClick={onClick}>
            <button className={classNames("btn btn-image btn-icon close-button")} onClick={onClose}>
                <img src={closeIconSvg} className={"removeIcon"} alt="Close icon for the full screen image lightbox" />
            </button>
        </div>
    );
}

export function Lightbox({ isOpen, onClose, children }: LightboxProps): ReactElement {
    const renderBackdropWithClose = useCallback(
        (props: RenderModalBackdropProps) => <BackdropWithClose onClose={onClose} {...props} />,
        [onClose]
    );

    return (
        <Modal
            show={isOpen}
            onHide={onClose}
            renderBackdrop={renderBackdropWithClose}
            className="mx-image-viewer-lightbox"
        >
            <div>{children}</div>
        </Modal>
    );
}
