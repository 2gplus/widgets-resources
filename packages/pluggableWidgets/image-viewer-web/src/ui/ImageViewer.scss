.mx-image-viewer {
    height: 100%;
    width: 100%;
}

.mx-image-viewer-responsive img {
    max-width: 100%;
    max-height: 100%;
}

.mx-image-viewer-lightbox-overlay {
    z-index: 1000;
    background-color: rgba(0, 0, 0, 0.85);
    position: fixed;
    inset: 0px;

    .close-button {
        position: absolute;
        top: 30px;
        right: 30px;
        color: white;
        display: flex;
        background: none;
        border: none;
        padding: 0;
        cursor: pointer;

        .removeIcon {
            height: 16px;
            width: 16px;
        }
    }
}
