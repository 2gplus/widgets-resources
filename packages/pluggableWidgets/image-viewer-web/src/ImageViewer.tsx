import { ValueStatus } from "mendix";
import { createElement, FunctionComponent } from "react";
import { ImageViewerContainerProps } from "../typings/ImageViewerProps";
import { ImageViewer as ImageViewerComponent } from "./components/ImageViewer";

function getImageUrl(props: ImageViewerContainerProps): string | undefined {
    switch (props.datasource) {
        case "dynamicImage":
        case "staticImage":
            return props.imageObject?.status === ValueStatus.Available ? props.imageObject.value.uri : undefined;
        case "imageUrl":
            return props.imageUrl?.status === ValueStatus.Available ? props.imageUrl.value : undefined;
        default:
            return undefined;
    }
}

export const ImageViewer: FunctionComponent<any> = (props: ImageViewerContainerProps) => {
    return (
        <ImageViewerComponent
            className={props.class}
            height={props.height}
            heightUnit={props.heightUnit}
            width={props.width}
            widthUnit={props.widthUnit}
            responsive={props.responsive}
            style={props.style}
            imageUrl={getImageUrl(props)}
        />
    );
};
