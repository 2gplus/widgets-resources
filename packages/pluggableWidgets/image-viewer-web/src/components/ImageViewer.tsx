import { createElement, CSSProperties, ReactElement, useCallback, useState } from "react";
import classNames from "classnames";
import Lightbox from "react-image-lightbox";
import { HeightUnitEnum, WidthUnitEnum } from "../../typings/ImageViewerProps";

import "../ui/ImageViewer.scss";
export interface ImageViewerProps {
    imageUrl: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
    className?: string;
    style?: CSSProperties;
    responsive: boolean;
    getRef?: (node: HTMLDivElement) => void;
    // TODO: Implement the Events
    onClick?: () => void;
}

function getStyle(value: string | number, type: string): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}

export function ImageViewer(props: ImageViewerProps): ReactElement | null {
    const [isOpen, setIsOpen] = useState<boolean>(false);

    const toggleLightBox = useCallback((): void => setIsOpen(currentIsOpen => !currentIsOpen), []);

    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.imageUrl }
            )}
            style={props.style}
        >
            <img
                onClick={toggleLightBox}
                src={props.imageUrl}
                style={{
                    ...props.style,
                    height: getStyle(props.height, props.heightUnit),
                    width: getStyle(props.width, props.widthUnit)
                }}
            ></img>
            {isOpen && props.imageUrl ? <Lightbox mainSrc={props.imageUrl} onCloseRequest={toggleLightBox} /> : null}
        </div>
    );
}
