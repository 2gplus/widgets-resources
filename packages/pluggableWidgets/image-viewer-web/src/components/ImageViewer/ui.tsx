import React, { createElement, CSSProperties, ReactElement } from "react";
import classNames from "classnames";
import { HeightUnitEnum, WidthUnitEnum } from "../../../typings/ImageViewerProps";
import { LightboxProps } from "../Lightbox";

import "../../ui/ImageViewer.scss";

function getStyle(value: string | number, type: WidthUnitEnum | HeightUnitEnum): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}
export interface ImageViewerWrapperProps {
    className?: string;
    responsive: boolean;
    hasImage: boolean;
    children:
        | ReactElement<ImageViewerGlyphicon | ImageViewerImage>
        | [ReactElement<ImageViewerGlyphicon | ImageViewerImage>, ReactElement<LightboxProps> | false];
}

export interface ImageViewerContentProps {
    style?: CSSProperties;
    onClick: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
}

function Wrapper(props: ImageViewerWrapperProps): ReactElement {
    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.hasImage }
            )}
        >
            {props.children}
        </div>
    );
}

export interface ImageViewerGlyphicon extends ImageViewerContentProps {
    icon: string | undefined;
    size: number;
}

function Glyphicon(props: ImageViewerGlyphicon): ReactElement {
    return (
        <span
            className={classNames("glyphicon", props.icon)}
            style={{ ...props.style, fontSize: `${props.size}px` }}
            onClick={props.onClick}
        />
    );
}

export interface ImageViewerImage extends ImageViewerContentProps {
    image: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
}

function Image(props: ImageViewerImage): ReactElement {
    return (
        <img
            src={props.image}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit)
            }}
            onClick={props.onClick}
        />
    );
}

export const ImageViewerUi = {
    Wrapper,
    Glyphicon,
    Image
};
