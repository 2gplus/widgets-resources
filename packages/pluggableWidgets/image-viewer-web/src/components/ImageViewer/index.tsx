import { createElement, CSSProperties, FunctionComponent, useCallback } from "react";
import { HeightUnitEnum, WidthUnitEnum, OnClickTypeEnum } from "../../../typings/ImageViewerProps";
import { useLightboxState } from "../../utils/lightboxState";
import { ImageViewerUi, ImageViewerContentProps } from "./ui";
import { Lightbox } from "../Lightbox";

import "../../ui/ImageViewer.scss";

export type ImageViewerImageProps = {
    type: "image" | "icon";
    image: string | undefined;
};

export interface ImageViewerProps extends ImageViewerImageProps {
    class: string;
    style?: CSSProperties;
    widthUnit: WidthUnitEnum;
    width: number;
    heightUnit: HeightUnitEnum;
    height: number;
    iconSize: number;
    responsive: boolean;
    onClickType: OnClickTypeEnum;
    onClick?: () => void;
}

export const ImageViewer: FunctionComponent<ImageViewerProps> = ({
    class: className,
    style,
    widthUnit,
    width,
    heightUnit,
    height,
    iconSize,
    responsive,
    onClickType,
    onClick,
    type,
    image
}) => {
    const { lightboxIsOpen, openLightbox, closeLightbox } = useLightboxState();

    const onImageClick = useCallback(() => {
        if (onClickType === "action") {
            onClick?.();
        } else if (onClickType === "enlarge") {
            openLightbox();
        }
    }, [onClick, onClickType, openLightbox]);

    const sharedContentProps: ImageViewerContentProps = {
        style,
        onClick: lightboxIsOpen ? undefined : onImageClick
    };

    const content =
        type === "image" ? (
            <ImageViewerUi.Image
                image={image}
                height={height}
                heightUnit={heightUnit}
                width={width}
                widthUnit={widthUnit}
                {...sharedContentProps}
            />
        ) : (
            <ImageViewerUi.Glyphicon icon={image} size={iconSize} {...sharedContentProps} />
        );

    return (
        <ImageViewerUi.Wrapper
            className={className}
            responsive={responsive}
            hasImage={image !== undefined && image.length > 0}
        >
            {content}
            {lightboxIsOpen && (
                <Lightbox isOpen={lightboxIsOpen} onClose={closeLightbox}>
                    {content}
                </Lightbox>
            )}
        </ImageViewerUi.Wrapper>
    );
};
