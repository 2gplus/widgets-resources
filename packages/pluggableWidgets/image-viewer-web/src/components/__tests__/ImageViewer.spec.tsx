import { createElement } from "react";
import { mount, ReactWrapper, shallow } from "enzyme";
import Lightbox from "react-image-lightbox";

import { ImageViewer, ImageViewerProps } from "../ImageViewer";

describe("ImageViewer", () => {
    const defaultProps: ImageViewerProps = {
        height: 300,
        heightUnit: "pixels",
        width: 300,
        widthUnit: "pixels",
        responsive: true,
        type: "image",
        image: "https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
    };

    const renderImageViewer = (props: ImageViewerProps): ReactWrapper<ImageViewerProps> =>
        mount(<ImageViewer {...props} />);

    it("renders the structure", () => {
        expect(shallow(<ImageViewer {...defaultProps} />)).toMatchSnapshot();
    });

    it("renders the lightbox when the image is clicked", () => {
        const imageViewer = renderImageViewer(defaultProps);

        const lightboxBefore = imageViewer.find(Lightbox);
        expect(lightboxBefore).toHaveLength(0);

        const image = imageViewer.find("img");
        expect(image).toHaveLength(1);

        image.simulate("click");

        const lightboxAfter = imageViewer.find(Lightbox);
        expect(lightboxAfter).toHaveLength(1);
        expect(lightboxAfter.prop("mainSrc")).toBe(defaultProps.image);
    });
});
