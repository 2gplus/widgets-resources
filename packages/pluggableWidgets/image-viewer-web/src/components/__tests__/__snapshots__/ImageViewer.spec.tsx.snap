// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageViewer renders the structure 1`] = `
<div
  className="mx-image-viewer mx-image-viewer-responsive"
>
  <img
    onClick={[Function]}
    src="https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
    style={
      Object {
        "height": 300,
        "width": 300,
      }
    }
  />
</div>
`;
